### PR TITLE
Ripple hashes integration

### PR DESCRIPTION
**Pull request by Stormtv**
According to #960 there was a need to remove the ripple hashes dependency and locally include it in this project.

I went through and copied the functionality of ripple hashes and rewrote the library in Typescript and implemented it and the unit tests associated with it.

Let me know if you have any questions or feedback.

Thanks!

### CODE REVIEW

**Comment by Stormtv** [comment]
One thing I removed from the ripple hashes lib was a dependency of https://github.com/crypto-browserify/createHash since this package only loads the crypto create hash package. I assume the existing webpack should browserify this without this package.

File: src/common/hashes/hashprefix.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// leaf node in tree
+export const HASH_LEAF_NODE = 0x4D4C4E00 // 'MLN'
+// inner transaction to sign
+export const HASH_TX_SIGN = 0x53545800 // 'STX'
+// inner transaction to sign (TESTNET)
+export const HASH_TX_SIGN_TESTNET = 0x73747800 // 'stx'
+// inner transaction to multisign
+export const HASH_TX_MULTISIGN = 0x534D5400 // 'SMT'
+
+export const HASH_LEDGER = 0x4C575200
```
Comment (keefertaylor) [inline src/common/hashes/hashprefix.ts]: Nit: document consistently with others

File: src/common/hashes/hashprefix.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * "space." This way, two different types of objects with the
+ * same binary data will produce different hashes.
+ *
+ * Each prefix is a 4-byte value with the last byte set to zero
+ * and the first three bytes formed from the ASCII equivalent of
+ * some arbitrary string. For example "TXN".
+ */
+
+// transaction plus signature to give transaction ID
+export const HASH_TX_ID = 0x54584E00 // 'TXN'
```
Comment (keefertaylor) [inline src/common/hashes/hashprefix.ts]: Any reason to not basket these up into an enum? 

It'll make your client code easier to read, I think:
`Hash.TX_ID' rather than `HASH_TX_ID` and imports will be cleaner

File: src/common/hashes/sha512hash.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import {createHash} from 'crypto'
+export default (hex: string): string => {
```
Comment (keefertaylor) [inline src/common/hashes/sha512hash.ts]: I personally find this syntax hard to parse at a glance. If you name the function rather than exporting something anonymous the code becomes more clear. 

Consider:
```
function sha512('hex string): string {
...
}

export default sha512;
```

File: src/common/hashes/shamap.ts [hunk 4]
```diff
@@ -0,0 +1,263 @@
+import {HASH_INNER_NODE, HASH_INNER_NODE_V2, HASH_TX_ID, HASH_TX_NODE, HASH_LEAF_NODE} from './hashprefix'
+import hash from './sha512hash'
+const HEX_ZERO = '0000000000000000000000000000000000000000000000000000000000000000'
+export const TYPE_INNER = 1
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: Enum?

File: src/common/hashes/shamap.ts [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+/**
+ * Inner (non-leaf) node in a SHAMap tree.
+ * @param {Number} depth (i.e. how many parent inner nodes)
+ * @class
+ */
+export class InnerNode extends Node {
+  public leaves: object
+  public type: 1 | 2 | 3 |4
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: These should be the type constants defined above

File: src/common/hashes/shamap.ts [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+  }
+}
+
+/**
+ * Inner (non-leaf) node in a SHAMap tree.
+ * @param {Number} depth (i.e. how many parent inner nodes)
+ * @class
+ */
+export class InnerNode extends Node {
+  public leaves: object
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: IIUC this is an object with keys that are characters 0-F? 

You could get better typing here if you change from 0-15 and then enforce that constraint I think?

File: src/common/hashes/shamap.ts [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+  public depth: number
+  public empty: boolean
+  public constructor(depth: number = 0) {
+    super()
+    this.leaves = {}
+    this.type = TYPE_INNER
+    this.depth = depth
+    this.empty = true
+  }
+  /**
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: nit: break between functions with vertical space (here and below)

File: src/common/hashes/shamap.ts [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+   * Overwrite the node that is currently in a given slot.
+   * @param {String} slot (a character 0-F)
+   * @param {Node} node (to place)
+   * @return {void}
+   */  
+  public set_node(slot: string, node: Node): void {
+    this.leaves[slot] = node
+    this.empty = false
+  }
+  public get_node(slot: string): Node {
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: can you doc this function? `slot`'s valid values are non-obvious.

File: src/common/hashes/shamap.ts [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+ * @param {Number} depth ()
+ * @class
+ */
+export class InnerNodeV2 extends InnerNode {
+  public common: string
+  public constructor(depth: number = 0) {
+    super(depth)
+    this.common = ''
+  }
+  /**
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: Add vertical space.

File: src/common/hashes/shamap.ts [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+      hex += this.leaves[slot] ? this.leaves[slot].hash : HEX_ZERO
+    }
+    const prefix = HASH_INNER_NODE.toString(16)
+    return hash(prefix + hex)
+  }  
+}
+
+/**
+ * V2 inner (non-leaf) node in a SHAMap tree.
+ * @param {Number} depth ()
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: OOC: I've never seen this style of doc. I usually document the constructors themselves. Is there a reason you prefer it?

File: src/common/hashes/shamap.ts [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+  public constructor(depth: number = 0) {
+    super()
+    this.leaves = {}
+    this.type = TYPE_INNER
+    this.depth = depth
+    this.empty = true
+  }
+  /**
+   * @param {String} tag (equates to a ledger entry `index`)
+   * @param {Node} node (to add)
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: OOC: why do you surround args in parenthesis?

File: src/common/hashes/shamap.ts [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+          this.set_node(tag[this.depth], new_inner_node)
+        }
+      }
+    } else {
+      // Neat, we have a nice open spot for the new node
+      this.set_node(tag[this.depth], node)
+    }
+  }
+  public get hash(): string {
+    if (this.empty) return HEX_ZERO
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: nit: stylistically, consider some vertical space to break up this function too.

File: src/common/hashes/shamap.ts [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+/**
+ * Leaf node in a SHAMap tree.
+ * @param {String} tag (equates to a ledger entry `index`)
+ * @param {String} data (hex of account state, transaction etc)
+ * @param {Number} type (one of TYPE_ACCOUNT_STATE, TYPE_TRANSACTION_MD etc)
+ * @class
+ */
+export class Leaf extends Node {
+  public tag: string
+  public type: 1 | 2 | 3 | 4
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: use type constants

File: src/offline/ledgerhash.ts [hunk 14]
```diff
@@ -1,5 +1,5 @@
 import * as _ from 'lodash'
-import hashes = require('ripple-hashes')
+import {computeLedgerHash as clh, computeTransactionTreeHash, computeStateTreeHash} from '../common/hashes'
```
Comment (keefertaylor) [inline src/offline/ledgerhash.ts]: I'd rather see us rename some functions than to alias a global import to an abbreviation of itself. 

Based on line 25, it looks like this function is actually `computeLedgerHeaderHash`, which I'd recommend renaming the exported function to.

File: test/hashes-test.js [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+      var sequence = 137;
+      var expectedEntryHash = '03F0AED09DEEE74CEF85CD57A0429D6113507CF759C597BABB4ADB752F734CE3';
+      var actualEntryHash = hashes.computeOrderHash(account, sequence);
+
+      assert.equal(actualEntryHash, expectedEntryHash);
+    });
+  });
+
+  describe('computeSignerListHash', function() {
+    /*
```
Comment (keefertaylor) [inline test/hashes-test.js]: I think we should probably omit this?

File: test/hashes-test.js [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+      var sequence = 82;
+      var expectedEntryHash = 'E35708503B3C3143FB522D749AAFCC296E8060F0FB371A9A56FAE0B1ED127366';
+      var actualEntryHash = hashes.computePaymentChannelHash(account, dstAccount, sequence);
+
+      assert.equal(actualEntryHash, expectedEntryHash);
+    });
+  });
+});
+
+// vim:sw=2:sts=2:ts=8:et
```
Comment (keefertaylor) [inline test/hashes-test.js]: remove

File: test/shamap-test.js [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+'use strict';
+
+var assert = require('assert');
+var SHAMap = require('../src/common/hashes/shamap').SHAMap;
+var TYPE_TRANSACTION_NM = require('../src/common/hashes/shamap').TYPE_TRANSACTION_NM
+
+var HEX_ZERO = '00000000000000000000000000000000' +
+               '00000000000000000000000000000000';
+
+function int_to_vuc(v) {
```
Comment (keefertaylor) [inline test/shamap-test.js]: convert this function, and the rest, to camel case to be consistent with the codebase.

Comment (keefertaylor) [inline test/shamap-test.js]: can you document this as well? I have no idea what `vuc` is

File: test/shamap-test.js [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+      ];
+
+      var shamapv2 = new SHAMap(2);
+      assert.equal(shamapv2.hash, HEX_ZERO);
+      fill_shamap_test(shamapv2, keys, hashesv2);
+    });
+  });
+});
+
+// vim:sw=2:sts=2:ts=8:et
```
Comment (keefertaylor) [inline test/shamap-test.js]: remove

File: src/common/hashes/shamap.ts (hunk 10 above)
Comment (Stormtv) [reply src/common/hashes/shamap.ts]: This was a remnant from the ripple-hashes documentation. I didn't modify the docs.

File: src/common/hashes/shamap.ts (hunk 11 above)
Comment (Stormtv) [reply src/common/hashes/shamap.ts]: This was a remnant from the ripple-hashes documentation. I didn't modify the docs.

File: src/common/hashes/shamap.ts (hunk 6 above)
Comment (Stormtv) [reply src/common/hashes/shamap.ts]: I'm not sure its a good idea to change the behavior of this but I did so in my last commit lmk what you think https://github.com/ripple/ripple-lib/pull/1039/commits/f63a364cd3e7668bbc6958b8a3051f136588b29e

File: src/common/hashes/shamap.ts [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+    let hex = ''
+    for (let i = 0; i < 16; i++) {
+      hex += this.leaves[i] ? this.leaves[i].hash : HEX_ZERO
+    }
+    const prefix = hashPrefix.INNER_NODE.toString(16)
+    return hash(prefix + hex)
+  } 
+}
+
+export class InnerNodeV2 extends InnerNode {
```
Comment (sublimator) [inline src/common/hashes/shamap.ts]: ShaMAP 2 has been canned

File: src/common/hashes/hash-prefix.ts [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+ * same binary data will produce different hashes.
+ *
+ * Each prefix is a 4-byte value with the last byte set to zero
+ * and the first three bytes formed from the ASCII equivalent of
+ * some arbitrary string. For example "TXN".
+ */
+
+ enum HashPrefix {
+  // transaction plus signature to give transaction ID
+  TX_ID = 0x54584E00, // 'TXN'
```
Comment (keefertaylor) [inline src/common/hashes/hash-prefix.ts]: A minor point, but this PR mixes abbreviations and words. Sometimes you refer to variables as `transaction` or `transactionHash` but you'll also refer to them as `tx` or `txHash`.  It would be great to standardize them. 

I'd lean towards not using abbreviations, but I don't have too strong of an opinion.

Comment (keefertaylor) [reply src/common/hashes/hash-prefix.ts]: This can be optional - I know that we're primarily just moving code here.

File: src/common/hashes/sha512hash.ts [hunk 21]
```diff
@@ -0,0 +1,7 @@
+import {createHash} from 'crypto'
+
+const hash = (hex: string): string => {
```
Comment (keefertaylor) [inline src/common/hashes/sha512hash.ts]: Optional: consider more specific naming, `sha512hash` or `sha512`

File: src/common/hashes/shamap.ts (hunk 19 above)
Comment (Stormtv) [reply src/common/hashes/shamap.ts]: Okay, I removed the support for ShaMAP 2

File: src/common/hashes/shamap.ts [hunk 22]
```diff
@@ -0,0 +1,288 @@
+import hashPrefix from './hash-prefix'
+import hash from './sha512hash'
+const HEX_ZERO = '0000000000000000000000000000000000000000000000000000000000000000'
+
+export enum NodeTypes {
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: nit: generally enums are named as a singular, `NodeType` vs `NodeTypes`.

This makes your code below more readable
```
// Sounds like this should be an array?  
public type: NodeTypes
```

changes to 
```
  // Ah, a singular enum value
  public type: NodeType
```

File: src/common/hashes/shamap.ts [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+  INNER = 1,
+  TRANSACTION_NM = 2,
```
Comment (keefertaylor) [inline src/common/hashes/shamap.ts]: I'd consider dropping abbreviations. It's not immediately obvious what `NM` or `MD` are.

Comment (keefertaylor) [reply src/common/hashes/shamap.ts]: @sublimator  Do you know what `MD` or `NM` stand for here?

Comment (sublimator) [reply src/common/hashes/shamap.ts]: NM = No metadata (used in consensus for proposed transaction sets iirc), and MD = transactions with metadata (as per a closed|validated ledger)

**Comment by keefertaylor** [comment]
Ship it!

**Comment by Stormtv** [comment]
Alright I think I've finished all the review changes lmk if there is anything else. If not this is good to go.

**Comment by sublimator** [comment]
Lol, seriously!??

**Comment by keefertaylor** [comment]
@sublimator Can you clarify what you think is still missing here?

**Comment by sublimator** [comment]
Haha - :)

**Comment by sublimator** [comment]
Just trolling :)
TBH I haven’t paid proper attention :/

**Comment by Stormtv** [comment]
@sublimator No rush on this. If you need more time to review it we can wait to merge this until your approval.

**Comment by sublimator** [comment]
Lol :)

**Comment by sublimator** [comment]
Sorry guys, old man blues :)

**Comment by intelliot** [comment]
Looks good to me, big thanks to all of you for the work and reviews! I went ahead and squashed this PR with the merge, hope that's OK. Let me know if anyone experiences any issues.
